data/blog/designing-sentrys-cross-region-replication.md

```diff
@@ -0,0 +1,160 @@
+    '/images/designing-sentrys-cross-region-replication/cdc-kafka.png',
+    '/images/designing-sentrys-cross-region-replication/outbox-sequence.png',
+  ]
+layout: PostLayout
+authors: ['markstory']
+---
+
+When we started designing how multi-region Sentry, we didn’t plan on having replication between regions. However, as we got farther in the design process it became clear that because of where our silo boundaries would be, we would need data replicated between regions to facilitate looking up in which region an organization was in, or validating API tokens. While these operations could be completed with Remote Procedure Calls (RPC). The latency, atomicity and resiliency impacts of these high-volume RPC operations wouldn’t be acceptable, and we needed a solution that would be more efficient and more correct.
+
+As the scenarios where we would use cross-region replication became more clear we collected the following requirements:
+
+- We needed resiliency to network failures. Replication would be disrupted by networking issues so we needed the system to work off of last-known information. During a disruption replicated data could become stale, but wouldn’t expire.
+- We wanted to be able to backlog changes without risk of an in-memory system overflowing and failing.
+- We wanted a solution that could reliably replicate changes eventually. Data loss should be rare, and we shouldn’t need to manually intervene to correct divergent state caused by networking or unavailability in other parts of the system.
+- We wanted a solution that we could gradually integrate into the application and validate replication both in CI, and in production before relying on it for customer traffic.
+- We didn’t want workloads from one customer to impact replication of another customer’s data.
+
+Our requirements led us to looking at solutions that would provide an eventually consistent model that leveraged durable data stores.
+
+## Scenarios where we use cross-region replication
+
+We use cross-region replication between our [region silos, and control silo](https://develop.sentry.dev/architecture/#silo-modes). Replication is performed in both directions but for different operations.
+
+For example, when an organization creates a new API token, that token needs to be present in both the control silo, and in the region where the organization is located.
+
+![outline of a example scenario for replication](/images/designing-sentrys-cross-region-replication/scenario-outline.png)
+
+Authentication tokens are centralized in Control Silo, so that we can organization tokens, work against control silo resources, and so that ensure that all tokens are unique with database constraints. We need to replicate the organization’s token to region silo so that it can be used to authenticate requests made there. Later when the token is deleted, the user will remove it from Control Silo, and replication should update the relevant region.
```

Comment:
Missing something here?

```suggestion
Authentication tokens are centralized in Control Silo, so that we can locate the organization tokens, work against control silo resources, and so ensure that all tokens are unique with database constraints. We need to replicate the organization’s token to region silo so that it can be used to authenticate requests made there. Later, when the token is deleted, the user will remove it from Control Silo, and replication should update the relevant region.
```